### PR TITLE
TT-72 when worker node in emergency mode, dont search keys in rpc

### PR DESCRIPTION
**Pull request by sredxny**
## Description
In the worker nodes, when they're in emergency mode then we should not try to search the keys in RPC in order to decrease the response time for those apis that are protected 

## Related Issue
Continuation of https://tyktech.atlassian.net/browse/TT-72

## Motivation and Context
Provide a full fix for https://tyktech.atlassian.net/browse/TT-72

## How This Has Been Tested
The same as https://github.com/TykTechnologies/tyk/pull/3163 but using an Api protected by key and another with Oauth

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Refactoring or add test (improvements in base code or adds test coverage to functionality)

## Checklist

- [x] Make sure you are requesting to **pull a topic/feature/bugfix branch** (right side). If pulling from your own
      fork, don't request your `master`!
- [x] Make sure you are making a pull request against the **`master` branch** (left side). Also, you should start
      *your branch* off *our latest `master`*.
- [ ] My change requires a change to the documentation.
  - [ ] If you've changed APIs, describe what needs to be updated in the documentation.
  - [ ] If new config option added, ensure that it can be set via ENV variable
- [ ] I have updated the documentation accordingly.
- [ ] Modules and vendor dependencies have been updated; run `go mod tidy && go mod vendor`
- [ ] When updating library version must provide reason/explanation for this update.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.
- [ ] Check your code additions will not fail linting checks:
  - [x] `go fmt -s`
  - [ ] `go vet`

### CODE REVIEW

**Comment by buger** [comment]
Any chance we can cover it with tests?

We have RPC mock (see the tests), and the test itself can look like: setting it to emergency mode, and testing if no requests were made to mock.

**Review by matiasinsaurralde — APPROVED** [review]
Looks good

**Review by matiasinsaurralde — APPROVED** [review]
As mentioned, would be great to add tests

**Comment by sredxny** [comment]
at this time all the grpc tests are broken, we need to fix them and then we can add a test for this ticket

**Review by matiasinsaurralde — APPROVED** [review]

**Review by matiasinsaurralde — APPROVED** [review]
This was already approved
@sredxny Could we fix the conflict and merge it?

File: gateway/middleware.go [hunk 1]
```diff
@@ -692,17 +692,20 @@ func (t BaseMiddleware) CheckSessionAndIdentityForValidKey(originalKey *string,
...
-	session, found = t.Spec.AuthManager.KeyAuthorised(key)
-	if !found && storage.TokenOrg(key) != t.Spec.OrgID {
-		//treat it as a custom key
-		key = generateToken(t.Spec.OrgID, key)
-		cacheKey = key
-		if t.Spec.GlobalConfig.HashKeys {
-			cacheKey = storage.HashStr(cacheKey)
-		}
+	// Only search in RPC if not in emergency mode
+	if !rpc.IsEmergencyMode() {
```
Comment (buger) [inline gateway/middleware.go:696]: You make the assumption here that `AuthManager` is RPC layer. 
I would rather make it explicit, like: 
if ok := t.Spec.AuthManager.(RPCConnection); ok && rpc.IsEmergencyMode() {
   return
}

File: gateway/rpc_test.go [hunk 2]
```diff
@@ -151,10 +152,19 @@ func TestSyncAPISpecsRPCFailure_CheckGlobals(t *testing.T) {
...
+	rpcMock := startRPCMock(dispatcher)
+	defer stopRPCMock(rpcMock)
+
+	store := RPCStorageHandler{}
+	store.Connect()
+	rpc.ForceConnected(t)
 
+	// Three cases: 1 API, 2 APIs and Malformed data
 	exp := []int{0, 1, 2, 2}
+	if *cli.HTTPProfile {
```
Comment (buger) [inline gateway/rpc_test.go:164]: What does this hack means?

**Comment by buger** [comment]
@sredxny pls add this test case to your new MDCB test framework.

File: gateway/middleware.go (hunk 1 above)
Comment (tbuchaillot) [reply gateway/middleware.go]: @buger modified. Let me know what you think 👍

File: gateway/middleware.go [hunk 3]
```diff
@@ -692,17 +692,20 @@ func (t BaseMiddleware) CheckSessionAndIdentityForValidKey(originalKey *string,
...
-	session, found = t.Spec.AuthManager.KeyAuthorised(key)
-	if !found && storage.TokenOrg(key) != t.Spec.OrgID {
-		//treat it as a custom key
-		key = generateToken(t.Spec.OrgID, key)
-		cacheKey = key
-		if t.Spec.GlobalConfig.HashKeys {
-			cacheKey = storage.HashStr(cacheKey)
-		}
+	// Only search in RPC if the storage is RPCStorageHandler and it's not in emergency mode
+	if _, ok := t.Spec.AuthManager.Store().(*RPCStorageHandler); ok && !rpc.IsEmergencyMode() {
```
Comment (buger) [inline gateway/middleware.go:696]: Now it will not run if Store is a different type :) 

You probably need smth like: !ok || !rpc.IsEmergencyMode()

**Review by tbuchaillot — APPROVED** [review]
LGTM

**Review by matiasinsaurralde — APPROVED** [review]

**Comment by sredxny** [comment]
cherry picked to release-3

**Comment by sredxny** [comment]
cherry picked to release-3-lts

**Pull request merged**